File: src/guard.ts
import { createHash, randomUUID } from 'node:crypto'
import { canonicalize } from './canonical.js'
import type { ActionReceipt, AgentAction, GuardResult, GuardRule } from './types.js'

/** The Action Receipt spec version this library implements. */
export const SPEC_VERSION = 'action-receipt/v0.1' as const

export interface GuardOptions {
  /** Recorded as agent_id on every receipt. */
  agent_id?: string
  /** Recorded as tenant_id on every receipt. */
  tenant_id?: string
  /**
   * Tamper-evident chaining. `true` gives the guard its own ReceiptChain;
   * pass a ReceiptChain instance to share one chain across components
   * (e.g. the pipeline links FINAL receipts after gate enrichment, so the
   * stored chain always verifies).
   */
  chain?: boolean | ReceiptChain
  /** Injectable clock, for deterministic tests. */
  now?: () => Date
}

/**
 * Links receipts into a tamper-evident hash chain. Each linked receipt
 * carries the SHA-256 of the previously linked one.
 *
 * IMPORTANT: link the receipt you intend to STORE. A receipt mutated after
 * linking will break verifyChain() — by design.
 */
export class ReceiptChain {
  private prev: string | undefined

  link(receipt: ActionReceipt): ActionReceipt {
    const linked: ActionReceipt =
      this.prev !== undefined ? { ...receipt, prev_receipt_hash: this.prev } : receipt
    this.prev = hashReceipt(linked)
    return linked
  }
}

/** A guard function: present an action, receive a decision and its receipt. */
export type Guard = (action: AgentAction) => GuardResult

function sha256(input: string): string {
  return createHash('sha256').update(input).digest('hex')
}

/** SHA-256 hex digest of the canonical JSON serialization of a payload. */
export function hashPayload(payload: Record<string, unknown>): string {
  return sha256(canonicalize(payload))
}

/** SHA-256 hex digest of a receipt's canonical JSON serialization. */
export function hashReceipt(receipt: ActionReceipt): string {
  return sha256(canonicalize(receipt))
}

/**
 * Creates a deterministic guard that evaluates agent actions against a
 * set of rules and issues an Action Receipt for every decision.
 *
 * Rules are pure predicates — no LLM is consulted, so the same action
 * always produces the same decision.
 *
 * @example
 * ```ts
 * const guard = createGuard([
 *   {
 *     id: 'refund-ceiling',
 *     check: (a) => a.type !== 'payment.refund' || (a.payload.amount_minor as number) <= 5000,
 *     reason: 'Refunds above £50.00 require human approval',
 *   },
 * ])
 *
 * const result = guard({ type: 'payment.refund', payload: { amount_minor: 9900 } })
 * // result.allowed === false; result.receipt is the audit record
 * ```
 */
export function createGuard(rules: GuardRule[], options: GuardOptions = {}): Guard {
  const chain =
    options.chain === true ? new ReceiptChain() : options.chain instanceof ReceiptChain ? options.chain : undefined

  return (action: AgentAction): GuardResult => {
    const failed = rules.filter((rule) => !rule.check(action))
    const allowed = failed.length === 0

    const receipt: ActionReceipt = {
      id: randomUUID(),
      spec: SPEC_VERSION,
      ...(options.agent_id !== undefined && { agent_id: options.agent_id }),
      ...(options.tenant_id !== undefined && { tenant_id: options.tenant_id }),
      action_type: action.type,
      payload_hash: hashPayload(action.payload),
      timestamp: (options.now?.() ?? new Date()).toISOString(),
      allowed,
      failed_rules: failed.map((rule) => rule.id),
      ...(allowed ? {} : { outcome: 'policy_blocked' as const }),
    }

    return {
      allowed,
      failed_rules: failed.map((rule) => rule.id),
      reasons: failed.map((rule) => rule.reason),
      receipt: chain ? chain.link(receipt) : receipt,
    }
  }
}

/**
 * Verifies a chained sequence of receipts: each receipt's
 * prev_receipt_hash must equal the hash of the receipt before it.
 * Returns the index of the first broken link, or -1 if the chain is intact.
 */
export function verifyChain(receipts: ActionReceipt[]): number {
  for (let i = 1; i < receipts.length; i++) {
    const expected = hashReceipt(receipts[i - 1]!)
    if (receipts[i]!.prev_receipt_hash !== expected) {
      return i
    }
  }
  return -1
}

File: python/suretyai/guard.py
from __future__ import annotations

import hashlib
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Callable, Optional

from suretyai.canonical import canonicalize
from suretyai.types import ActionReceipt, AgentAction, GuardResult, GuardRule

SPEC_VERSION = "action-receipt/v0.1"

Guard = Callable[[AgentAction], GuardResult]


@dataclass
class GuardOptions:
    agent_id: Optional[str] = None
    """Recorded as agent_id on every receipt."""

    tenant_id: Optional[str] = None
    """Recorded as tenant_id on every receipt."""

    chain: bool = False
    """When True, each receipt carries the SHA-256 hash of the previous one."""

    now: Optional[Callable[[], datetime]] = None
    """Injectable clock, for deterministic tests."""


def _sha256(data: str) -> str:
    return hashlib.sha256(data.encode()).hexdigest()


def hash_payload(payload: dict) -> str:  # type: ignore[type-arg]
    """SHA-256 hex digest of the canonical JSON serialization of a payload."""
    return _sha256(canonicalize(payload))


def hash_receipt(receipt: ActionReceipt) -> str:
    """SHA-256 hex digest of the canonical JSON serialization of a receipt."""
    d = {k: v for k, v in receipt.__dict__.items() if v is not None}
    return _sha256(canonicalize(d))


def create_guard(rules: list[GuardRule], options: Optional[GuardOptions] = None) -> Guard:
    """Create a deterministic guard that evaluates agent actions against a set of rules.

    Rules are pure predicates — no LLM is consulted, so the same action
    always produces the same decision.

    Example::

        guard = create_guard([
            GuardRule(
                id="refund-ceiling",
                check=lambda a: a.type != "payment.refund" or a.payload["amount_minor"] <= 5000,
                reason="Refunds above £50.00 require human approval",
            )
        ])

        result = guard(AgentAction(type="payment.refund", payload={"amount_minor": 9900}))
        assert not result.allowed
    """
    opts = options or GuardOptions()
    prev_receipt_hash: Optional[str] = None

    def guard(action: AgentAction) -> GuardResult:
        nonlocal prev_receipt_hash

        failed = [rule for rule in rules if not rule.check(action)]
        allowed = len(failed) == 0

        now_dt = opts.now() if opts.now else datetime.now(tz=timezone.utc)

        receipt = ActionReceipt(
            id=str(uuid.uuid4()),
            spec=SPEC_VERSION,
            agent_id=opts.agent_id,
            tenant_id=opts.tenant_id,
            action_type=action.type,
            payload_hash=hash_payload(action.payload),
            timestamp=now_dt.isoformat(),
            allowed=allowed,
            failed_rules=[r.id for r in failed],
            outcome=None if allowed else "policy_blocked",
            prev_receipt_hash=prev_receipt_hash if opts.chain else None,
        )

        if opts.chain:
            prev_receipt_hash = hash_receipt(receipt)

        return GuardResult(
            allowed=allowed,
            failed_rules=[r.id for r in failed],
            reasons=[r.reason for r in failed],
            receipt=receipt,
        )

    return guard


def verify_chain(receipts: list[ActionReceipt]) -> int:
    """Verify a chained sequence of receipts.

    Returns the index of the first broken link, or -1 if the chain is intact.
    """
    for i in range(1, len(receipts)):
        expected = hash_receipt(receipts[i - 1])
        if receipts[i].prev_receipt_hash != expected:
            return i
    return -1
